src/util/alloc/allocator.rs

```diff
@@ -102,32 +116,70 @@ pub fn get_maximum_aligned_size<VM: VMBinding>(
     }
 }
 
+/// A trait which implements allocation routines. Every allocator needs to implements this trait.
 pub trait Allocator<VM: VMBinding>: Downcast {
+    /// Return the [`VMThread`] associated with this allocator instance.
     fn get_tls(&self) -> VMThread;
 
+    /// Return the [`Space`] instance associated with this allocator instance.
     fn get_space(&self) -> &'static dyn Space<VM>;
+
+    /// Return the [`Plan`] instance that this allocator instance is associated with.
     fn get_plan(&self) -> &'static dyn Plan<VM = VM>;
 
-    /// Does this allocator do thread local allocation? If an allocator does not do thread local allocation,
-    /// each allocation will go to slowpath and will have a check for GC polls.
+    /// Return if this allocator can do thread local allocation. If an allocator does not do thread
+    /// local allocation, each allocation will go to slowpath and will have a check for GC polls.
     fn does_thread_local_allocation(&self) -> bool;
 
-    /// At which granularity the allocator acquires memory from the global space and use them as thread local buffer.
-    /// For example, bump pointer allocator acquire memory at 32KB blocks. Depending on the actual size for the current object,
-    /// they always acquire memory of N*32KB (N>=1). Thus bump pointer allocator returns 32KB for this method.
-    /// Only allocators that do thread local allocation need to implement this method.
+    /// Return at which granularity the allocator acquires memory from the global space and use
+    /// them as thread local buffer. For example, the [`BumpAllocator`] acquires memory at 32KB
+    /// blocks. Depending on the actual size for the current object, they always acquire memory of
+    /// N*32KB (N>=1). Thus the [`BumpAllocator`] returns 32KB for this method.  Only allocators
+    /// that do thread local allocation need to implement this method.
     fn get_thread_local_buffer_granularity(&self) -> usize {
         assert!(self.does_thread_local_allocation(), "An allocator that does not thread local allocation does not have a buffer granularity.");
         unimplemented!()
     }
 
+    /// An allocation attempt. The implementation of this function depends on the allocator used.
+    /// If an allocator supports thread local allocations, then the allocation will be serviced
+    /// from its TLAB, otherwise it will default to using the slowpath, i.e. [`alloc_slow`].
+    ///
+    /// Note that in the case where the VM is out of memory, we return a null pointer back to the
+    /// VM and expect that the OOM error is thrown by the binding.
+    ///
+    /// Arguments:
+    /// * `size`: the allocation size in bytes.
+    /// * `align`: the required alignment in bytes.
+    /// * `offset` the required offset in bytes.
     fn alloc(&mut self, size: usize, align: usize, offset: isize) -> Address;
 
+    /// Slowpath allocation attempt. We break up the slowpath implementation into two parts
```

Comment:
I don't know what are the two parts of our slowpath implementation.

Reply:
Yes, I agree. "Parts" is misleading. I mean to say different functions since one is `#[inline(never)]` for cache considerations (to the best of my understanding).

Reply:
Do you have a suggestion as to what terminology to use? I think using "functions" would work. Alternatively, I can remove the sentence entirely as they may confuse the reader (although it would be good to document optimizations if we can).

Reply:
If you are talking about `alloc_slow()` vs `alloc_slow_inline()` vs `alloc_slow_once()`, this is my guess, after checking the Java MMTk:
* `alloc_slow()` is the slowpath function that usually will be used. It is explicitly out of line for performance. It may be overriden by some allocators and implement their own slowpath semantics (e.g. bump pointer allocator in Java MMTk overrides this method and implements card marking).
* `allow_slow_once()` is a single slowpath allocation attempt. Every allocator needs to implement it.
* `alloc_slow_inline()` has the logic of retrying slowpath allocation and is responsible of dealing with OOM. We do not expect any allocator to override it. The method exists because some allocators like the immix allocator use an intermediate path (something between fastpath and slowpath), and call this in the intermediate path. By default, `alloc_slow()` just calls `alloc_slow_inline()`.